kyo-data/shared/src/main/scala/kyo/Result.scala

```diff
@@ -393,183 +382,181 @@ object Result:
+        inline def mapFailure[E2](inline f: E => E2): Result[E2, A] =
+            foldAll(Result.panic)(e => Result.fail(f(e)))(Result.succeed)
 
-        /** Applies a function to the error value of this Result.
+        /** Maps only the panic exception to an error value.
           *
           * @param f
-          *   The function to apply
-          * @tparam E2
-          *   The type of the resulting error
+          *   The function to apply to the panic exception
           * @return
-          *   A new Result after applying the function to the error
+          *   A new Result with the mapped panic as a failure
           */
-        inline def mapFail[E2](inline f: E => E2): Result[E2, A] =
-            self match
-                case Fail(e) =>
-                    try Fail(f(e))
-                    catch
-                        case ex => Panic(ex)
-                case _ => self.asInstanceOf[Result[E2, A]]
+        inline def mapPanic[E2](inline f: Throwable => E2): Result[E | E2, A] =
+            try foldAll(e => Result.fail(f(e)))(Result.fail)(Result.succeed)
+            catch
+                case ex => Panic(ex)
 
-        /** Applies a predicate to the successful value of this Result.
+        /** Applies a function to the successful value of this Result.
           *
-          * @param p
-          *   The predicate to apply
+          * @param f
+          *   The function to apply
           * @return
-          *   A new Result that fails if the predicate doesn't hold
+          *   A new Result after applying the function
           */
-        inline def withFilter(inline p: A => Boolean): Result[E | NoSuchElementException, A] =
-            filter(p)
+        inline def flatMap[E2, B](inline f: A => Result[E2, B]): Result[E | E2, B] =
+            self match
+                case self: Error[E] @unchecked => self
+                case self =>
+                    try f(self.asInstanceOf[Success[A]].getOrThrow)
+                    catch
+                        case ex =>
+                            Panic(ex)
 
-        /** Applies a predicate to the successful value of this Result.
+        /** Handles specific error types with a recovery function.
           *
-          * @param p
-          *   The predicate to apply
+          * @param f
+          *   The function to apply to matching errors
+          * @tparam E2
+          *   The specific error type to handle
           * @return
-          *   A new Result that fails if the predicate doesn't hold
+          *   A new Result after applying the recovery function
           */
-        inline def filter(inline p: A => Boolean): Result[E | NoSuchElementException, A] =
-            flatMap { v =>
-                if !p(v) then
-                    Fail(new NoSuchElementException("Predicate does not hold for " + v))
-                else
-                    v
-            }
+        inline def flatMapError[E2 <: E](
```

Comment:
These APIs are quite powerful given that they can transform any of the states and freely handle errors